source-git: provide E2E guide (#231)
packit/packit.dev: content/docs/source-git/work-with-source-git/_index.md

---
title: "Working with source-git"
date: 2020-06-24T23:59:59+01:00
draft: false
weight: 3
bookCollapseSection: true
---

# Working with source-git

This document is focused on how packit treats a source-git repo.

> If you are looking for the design concept of source-git, please head on here:
> [docs/source-git-design]({{< ref "/docs/source-git/design.md" >}})

Reviewer: ```suggestion
If you are looking for the design concept of source-git, please [read it here]({{< ref "/docs/source-git/design.md" >}}).
```

The source-git concept is fairly simple - take an upstream repo and
shove downstream content on top of it. Easy right?

Packit is then able to work with such a repo and:

* Create SRPMs

* Sync the state into dist-git

* And you are able to utilize all features of packit-service: RPM builds of PRs,
  tests on those builds, and getting feedback in pull (merge) requests.

Let's describe all of these in detail.

## Upstream git history or just an unpacked tarball?

Your source-git repo can contain upstream git history if you want. Or it

> doesn't. The alternative is to unpack a tarball corresponding to an upstream

Reviewer: Wording feels sloppy, could you make it a little bit more concrete and structured?

Would it make sense to reference docs for creating such source-git repositories?

Author: I am going to drop that section for good, you're right that it doesn't feel right to have it here.

> release. This is really up to you and packit is able to work with both.

Reviewer: Do we want to promote this alternative?
How the history is going to look like. Is a source-git repo like this still useful for development? Does it considerably easier to maintain the package?

The key ingredient here is the config option `upstream_ref`. It tells packit
the git ref which is supposed to match the archive specified in the spec file.
All commits on top of this ref are treated either as:

* downstream patches
* downstream packaging

For downstream patches, packit creates a patch file for each commit on top of
`upstream_ref`. Downstream packaging changes are ignored in patch files since
they can be directly committed to dist-git. You can control the ignore

> mechanism with config option `patch_generation_ignore_paths`.

Reviewer: Link the docs of this config option.

### Example

Let's have a look at source-git repo for pacemaker package:

    $ git log --online
    bd12722 (HEAD -> c8s-test, origin/c8s-test, master) downstream packaging
    7201e28 Refactor: controller: remove unused argument
    91c557c Refactor: controller: convert active_op_t booleans to bitmask
    2e90063 Refactor: controller: rename struct recurring_op_s to active_op_t
    3ee90ce (tag: source-git-start) pacemaker-2.0.3 base

We have 5 commits:
* bd12722 - top commit, contains a spec file so packit won't create a patch file out of this one
* 7201e28 - the 3rd downstream patch
* 91c557c - the 2nd downstream patch
* 2e90063 - the 1st downstream patch
* 3ee90ce - the bottom commit, equals to an unpacked upstream tarball

packit.yaml has only two lines:

    $ cat .packit.yaml
    specfile_path: SPECS/pacemaker.spec
    upstream_ref: source-git-start

As you can see, we are telling packit, that commit tagged as `source-git-start`

> indicates the barrier between upstream and downstream, in our case, it's just a

Reviewer: Not sure this belongs here, but couldn't this start be indicated in some other way? Cannot this be the name of the upstream branch, for example? Where I would like to get with this: could we remove the requirement of creating a tag to mark the border between upstream and downstream and use, for example, only the name of the upstream branch?

Author: This is an old example which I did for Stream 8 and just moved it here, I'm gonna redo this section using a real upstream tag, that's a good point.

single commit with the upstream archive content, but it could have been the
whole upstream git history.

### Rebase or merge?

> A common workflow in the world of open source development is to accept

Reviewer: This sounds very fancy but it's very low on information :)

Could it be simplified?

Author: will add links, sounds like an overkill to explain pull request workflow here

pull requests and have merge commits in the git history. If the PRs are not

> rebased against the main branch before the merge, the history may go wild.

Reviewer: Could "go wild" replaced by a more technical/concrete expression?

Author: yup, made it more accurate

Packit is able to work with such a state by creating an ephemeral branch with
linear history and generate the patch files out of it. [More

> info](https://github.com/packit/packit/pull/766).

Reviewer: Shouldn't we write this up and create and article for it, instead of referencing a PR? Seeing a PR might scare readers away.

Author: Since this is very technical, I'm personally fine with pointing to a PR (instead of not describing it at all) - are we able to cover this by an article? Would anyone read it or find it useful?

Picking up latest upstream changes into your downstream source-git repo has
multiple solutions and it's up to you to pick the one which suits your workflow
best:

1. **Rebase against the main upstream branch** (a.k.a. `git pull --rebase
   upstream master`). This solution implies that you are able to perform the
   force-push operation (e.g. Fedora and CentOS dist-git instances don't allow
   force-pushes). Packit will not help you with the rebase process in any way
   and it's up to you to complete the rebase.
2. **Create a new branch for every upstream release**. No rebase needed — may
   require good branch hygiene since the number of branches can skyrocket. No

>    dedicated help from packit again, you are running all the `git` commands.

Reviewer: Wondering if there would be any "hand-holding" that Packit could do here? :thinking:

Author: We could certainly create an automation for this, that would be pretty awesome.
